### ISSUE TITLE
Rethink "production builds" and get rid of production profile

### ISSUE THREAD

**Issue by mstahv** [vaadin/flow#17737]
### Describe your motivation

Currently our project build files looks ugly and you need special Vaadin skills to build a deployable artifact (typically in Maven project build with `production` profile that builds the front-end bundle). This makes deploying Vaadin applications harder than it should be and requires special tricks when for example using packeto buildpacks to pack your Vaadin app. 

In addition to "just" not being like a standard Java project, this makes Vaadin build files look much more complicated.

### Describe the solution you'd like

Make our Maven plugin smarter so that we don't need a special profile for it. It could detect if the front-end build is needed from the environment and/or have some flag to trigger developmentmode/productionmode.

With cloud native buildpacks I have for example used this configuration in my pom.xml to enable production front-end build:

            <activation>
                <property>
                    <name>env.CNB_TARGET_OS</name>
                    <value>linux</value>
                </property>
            </activation>

But there are probably other ways to detect the situation as well. And most likely we need a flag to set this explicitly as well.

**Comment by knoobie** [comment]
For Jenkins I do: 

```xml
   <activation>
        <!-- Profile that is automatically activated when run in Jenkins because Jenkins creates a env.BUILD_NUMBER-->
        <property>
          <name>env.BUILD_NUMBER</name>
        </property>
    </activation>


```

**Comment by Legioth** [comment]
Could we do production builds by default? If you do `mvn` or `mvn spring-boot:run`, then you want to run in development mode. If you do `mvn package` or `mvn install`, then you want to build for production. The same is also the case with Quarkus or if you use e.g. `wildfly-maven-plugin` or `jetty-maven-plugin`.

It's a special case if you want to build a .war or .jar that _isn't_ for production use. Maybe we should have a dedicated profile (or just documentation) for that case instead?

**Comment by knoobie** [comment]
I'm not sure how good the automatic detection works with all the possible configuration options for the flow maven plugin and e.g. looking at my configuration how different it is for development and production mode. I also often run mvn clean install locally in multi module projects where Vaadin is only used in one and I don't care about a runnable artefact for that anyway, cause I'm starting it locally with spring boot anyway and just need to refresh IDEAs cache or testing other things.

**Comment by Legioth** [comment]
Good point. You might do `mvn install` locally because you've got a multi-module project even though you're not actually interested in a slow production build for for the Vaadin part.

On the other hand, we might also have a simpler default for tutorials and starters that are based on the single-module use case where production could be the default and then separately show a multi-module setup where a production build isn't the default?

**Comment by knoobie** [comment]
It depends on the audience Vaadin wants to target with this. I don't have the numbers, but I would personally think that multi module projects are wider used in corporate environments, so this change would only benefit starters in terms of repo and developer / user. 

I personally would more focus on the idea of Matti and improve auto detection of environments where the production build normally runs (GitHub, gitlab, Jenkins, bit bucket and so on) and activate the profile by default

**Comment by mstahv** [comment]
The most common case for development mode is you launch it via IDE's `public static void main string args` support. If we can detect that (or everything else) and provide a flag to set development mode explicitly 🤔

**Comment by knoobie** [comment]
Now I'm feeling like a unicorn, never done that, always using the mvn boot run command 🤫

**Comment by mcollovati** [comment]
> Now I'm feeling like a unicorn, never done that, always using the mvn boot run command

You're not alone, we are at least in two :grin:

**Comment by Legioth** [comment]
One thing to keep in mind here is that we're always balancing between two partially conflicting goals:
1. Keep the getting-started experience as easy as possible for new users. We want to make sure they immediately get on a path towards succeeding with Vaadin. This means that we want to have good defaults and we want to make things like the default `pom.xml` file look simple rather than intimidating.
2. Help users succeed with all the complex stuff that pops up in real-world applications. It's acceptable if users need to do som additional setup based on their needs as long as we can guide them in the right direction.

**Comment by knoobie** [comment]
Don't get me wrong, I totally understand the desire and need to appeal to new developer - especially those that are scared away by more than two lines of config (over exaggerated). But I feel that the amount of "goodies" are already plenty out there with the default development and prod bundle making the whole development process for new developer or rapid prototyping relatively painless. 

Just some ideas: 
- add more comments to the pom with explanations about each important part with links to the docs for further reading 
- add a do 1., 2. and 3. depending on your needs to the pom's comment on top so that everyone can read it 
- Improve the workflow for new users that development and prod bundle can work *without* the need of rebuilding even when adding css (e.g. serve those as static resources) so that the whole additional config for production is gone / can be a text / explanation in the pom when people need it.. or even: 
- add a prominent message to the log: "you have added an addon / Javascript which needs a new development / prod bundle: read here more about it. You can also find here ready made snippets for your pom" 
- Make flow more intelligent in guiding users: https://github.com/vaadin/flow/pull/17745#discussion_r1341376690

**Comment by mstahv** [comment]
Here is a design for you. Prototyped this a bit with V24.4 test project and at least for Spring Boot based projects I'm quite confiden this would work pretty well:

 * vaadin-dev artifact should be made optional in core artifacts and [defined as optional](https://gist.github.com/mstahv/42b041a74ae4b5abcf58e9257fbc1e53#file-pom-xml-L80-L88) in projects, just like spring-boot-devtools. This way no ugly exclusions are needed in production profile, it is simply left out of the production artifact. This could be done eagerly without big regression risks even in 24.4 🤔 Release notes & clear error message would make it easy for developers to "add back" the vaadin-dev as optional dependency.  Now that there are also other dependency dance, it would be a perfect time to do this kind of change.
 * The Vaadin maven plugin should work like spring boot plugin, its magic should be enabled implicitly. E.g. skip goals if spring-boot:run is executed by @knoobie via CLI/IDE 😜 Add flags to enable/disable explicitly for thing that are needed.
 * Without special configuration, maven plugin generated stuff should be ignored if vaadin-dev is on the classpath. Basiclly we would by default start in dev mode, even if "production front-end stuff" is available.
 * The production profile should be entirely removed from new starters.
 
 In the [prototype](https://gist.github.com/mstahv/42b041a74ae4b5abcf58e9257fbc1e53):
 
  * production build, ok and "cloud ready" (build pack makes a production ready artifacts without further dance 💪)
  * development mode ok , unless run with spring-boot:run (which triggers vaadin maven plugin as it doesn't have the "magic" and is configured to do the "production build" for better defaults).
  * Various required exculsions (because dev-mode is direct dependency currently) make the build look almost as ugly as the current default starter 🤪
  
I suggest to do the dev-mod scope change "today", for Vaadin 24.4, and others ASAP without a rush (at least with a bit of creativity I believe it to be possible to introduce these in a perfectly backwards compatible manner whenever we want, only build plugin changes required).

**Comment by shymega** [comment]
I'm just curious about this sentence from OP's post:

>  requires special tricks when for example using Packeto buildpacks to pack your Vaadin app.

I'm trying to deploy a Hilla app on Fly.io, with Spring Native and Paketo Buildpacks. However, I can't get past the `build-frontend` exception due to Vite not being located.

I'm assuming OP was referring to that. I would be very grateful if the workaround for this was put in the original issue, or as a comment - just as a workaround. I've subscribed to this issue in the meantime.

**Comment by knoobie** [comment]
Please create a issue in Hilla; it should work if configured correctly and used. This issue is about resucing the need for special config altogether.

**Comment by shymega** [comment]
> Please create a issue in Hilla; it should work if configured correctly and used. This issue is about resucing the need for special config altogether.

Apologies, I felt this is also a Flow issue as the problematic Maven plugin [inherits](https://github.com/vaadin/hilla/blob/main/packages/java/maven-plugin/src/main/java/com/vaadin/hilla/maven/BuildFrontendMojo.java#L36) from Flow.

Feel free to hide my original comment as off-topic, and I'll do some more digging to figure out where the best place for the issue would be. At this point, I'm fairly certain it's a Flow Maven issue.

I'm still curious about the workaround OP used (if they did?), but I appreciate this isn't the place to discuss that.

**Comment by mstahv** [comment]
Sorry I missed the continued discussion here and returning to the subject bit late (I have overflow in GH notifications 😬). Did you try the trick/workkaround I use (mentioned in the issue description) to active the production profile by default? I think that should work for you.

**Comment by mstahv** [comment]
I have been iterating a PoC towards my understanding of "the ideal state", back and forth, with different approaches, for way too long without releasing anything for public discussion. Just a while ago I finally pushed our now a first public iteration for comments. Maven central is probably still syncing, but quite soon you ought to be able to test the example project from this repository https://github.com/viritin/viritin-sb

Things to notice and test, the important parts:

 * No special profiles you need to learn about
 * mvn install creates a production ready artifact, no special tricks or Vaadin knowledge required typical cloud deployments or CI/CD pipelines
 * Starting the app from IDE (main method) ( or in the @knoobie method 😉 using mvn spring-boot:run), gets you into development mode (no special Vaadin knowledge needed if you want back to the dev mode after mvn package)
 * the pom.xml (of the example project) is short and clean

With these things in core Vaadin would feel much easier to take into use for experienced Java developer (partly perceived feeling, but very important!) and would provide less head-ache when integrating to various systems in the cloud era.
 
 Notes about implementation details of this poc in its current form:

 * The PoC is currently Spring Boot only
 * The Hilla setup is currently broken, but probably fixable easily for somebody is familiar with the module structure there
 * The PoC is bit fragile for version changes, if we change what is needed by vaadin-dev (which is a LOT these days). The dev mode exclusion is currently hard coded, as I got frustrated fiddling with my custom maven plugin that would have detected the correct transitive dependency set and would have fed that for spring-boot-maven-plugin as artifacts to exclude from the "repackaged" artifact.
 * The final solution can naturally be built in a different way if bolted directly to core Vaadin artifacts
 * Some more advanced Maven experts can probably come up with better solutions to accomplish the same end user value as this kind of poc on top of current Vaadin release. I'm very eagerly waiting for contributions or alternative solutions.
 *  Probably should have built the PoC with Gradle as it is easier platform for various hacks.
 
Heading now to test this in a couple of hobby projects, please help my by doing the same!

**Comment by mstahv** [comment]
Update, version 0.0.2 of the PoC published in previous comment also works with Hilla (smoke tested).

**Comment by mstahv** [comment]
Updates regarding my PoC:

 * Version 0.0.3, now uses quite different approach to get pretty much the same DX: The default scope of dependencies is now targeted for production and 'test' scope contains the "dev mode stuff". Kind of what Leif suggested earlier, but instead of custom profile for development, utilising `test` scope.
   * Same approach as suggested by Spring Boot to use Testcontainers in development time.
   * Requires another Spring Boot main method to `src/test/java` (bit more code, although can be handy for other test time configuration and e.g. Testcontainers) usage
 * There is also a "viritin-addon-project-parent" for Vaadin add-ons (no need to be a Viritin add-on 🤓). Much cleaner build possible for add-ons as well and Spring Boot works great as a test server (I have used this approach in more complex form in other add-ons as well. Starts faster than Jetty and provides features one might want to utise in tests.

**Comment by Dudeplayz** [comment]
> * Starting the app from IDE (main method) ( or in the @knoobie method 😉 using mvn spring-boot:run), gets you into development mode (no special Vaadin knowledge needed if you want back to the dev mode after mvn package)

Please keep quarkus:dev in mind, as it is clear for what purpose it is used. [Quarkus also allows remote development](https://quarkus.io/guides/maven-tooling#remote-development-mode), which could be a use case were an build is intended with development bundle. I am not using this actually, but it came in my mind.

**Comment by mstahv** [comment]
Good point @Dudeplayz. I'm quite confident that this approach can be applied in some for Quarkus apps as well at least with some plugin code. As dev mode is in quarkus started with the CLI tooling or maven plugin, it might even be easier to intercept dev mode startups. Maybe @mcollovati has some good ideas how to get a similar exprience with Quarkus🤔

**Comment by ThomasVitale** [comment]
@mstahv I've been facing similar issues with some of my Spring Boot libraries and raised a couple of feature requests to the Spring Boot project to address them. I wonder if any of that could be useful to Vaadin and if you had any input to share:

* https://github.com/spring-projects/spring-boot/issues/46368
* https://github.com/spring-projects/spring-boot/issues/46817

**Comment by mstahv** [comment]
Thanks @ThomasVitale 👍 Something is indeed missing from SB and maybe I'd claim that more wider level, from IDE and build tools. After my endevours on this area, and even making one Spring Boot Maven Plugin PR that I hoped to open door for better solutions, I feel that hacking around those with something like Spring Boot Maven/Grade plugin is kind of half backed solution.

I have been very happy lately with the approach I have in the "viritin-sb" PoC currently, which is essentially the same Spring Boot takes with [TestContainers for development time](https://docs.spring.io/spring-boot/reference/features/dev-services.html#features.dev-services.testcontainers). So practically bit "abusing" the "test" scope (which is available in both Maven and Gradle and supported by IDEs) and then having another "main method" for development time that automatically wires things needed during development. Although this issue seems like it is stalled, I have been using that in my own apps and demos lately and I'm very happy for the exprience both when developing and deploying. Also no ugly looking exclusions and speciap profiles needed in the build file.

In all my recent projects/demos where I have HAD to work with RDBMS, I have used Testcontainers in the same way too. In this kind of cases there is two "techs" that benefit from the same "DevModeApplication" class.

The cons:

 * Extra main method/class (~ DevModeApplication) needed in src/test folder
 * People are not familiar with the approach, even it is now promoted with TestContainers. I actually [suggested a change to hint](https://github.com/spring-projects/spring-boot/pull/43571) about possible "DemoModeApplication" available to Spring Boot, but @philwebb shot it down because it was not "perfect". I didn't fully understand how much more perfect the current "embedded DB hint" is, which kind of feels to me that shouldn't even exist in the era of containers, but somehow it sounded like they have some bigger idea how to solve these issues in a different way 🤷‍♂️

You should probably also follow SB issue https://github.com/spring-projects/spring-boot/issues/13289, which I think is almost like bug than a feature (that others but spring boot dev tools are not excluded from uber jar packaging). Getting that done might open doors to solve these dev time helpers in other way as well.

**Comment by mstahv** [comment]
Oh, looks like @snicoll already pointed you to https://github.com/spring-projects/spring-boot/issues/13289 from your second issue.

**Comment by mshabarov** [comment]
**Acceptance Criteria** suggestions that I could draft after giving a try to prototype and Vaadin Quarkus add-on starter:

- [ ] **Production builds**
   Any Vaadin application must be buildable for production by running `mvn package`. The build should produce a production-ready artifact without requiring additional manual steps or profile.

- [ ] **Development Mode: Spring Boot**
   For Spring Boot applications, developers can: copy the `Application` class into the test package and run in dev mode using `mvn spring-boot:test-run` or by using the IDE “Run/Play” button. The dev-mode runtime must function equivalently to running the main `Application` class, if `Application` and `DevModeApplication` have the same codes (annotations/parent classes/interfaces). `DevModeApplication` can also define extra test dependencies like test repositories/data sources/beans thanks to Spring's _TestContainers_ feature.

- [ ] **Development Mode: Quarkus**
   For Quarkus applications, developers must be able to run in dev mode via either: the `quarkus:dev` maven goal, or a `QuarkusMain` annotated main class that, with some customisations, can pick the platform's `vaadin-dev` with a test scope and run Vaadin in dev mode.

- [ ] **Development Mode: Spring Framework / Plain Servlet**
   For Spring Framework or plain servlet applications, it is possible to define a dev profile in the build configuration (e.g., pom.xml or build.gradle) that adds the `vaadin-dev` dependency with runtime scope and then run the application in dev mode using that profile.

- [ ] **Switching between modes**
   Vaadin cleans up build data when switching from production builds and dev runs (and vice versa), ensuring no runtime errors occur due to inconsistent state/configuration. Specifically, no scenario where `vaadin-dev` artifact remain while the application is running in production mode.

- [ ] **Autodetect build environments**
   As a secondary helper feature, Vaadin could automatically detects CI/CD build environments that containerise the app for production use and run the production build, e.g. cloud native buildpacks, Heroku/Cloud Foundry, Kubernetes/OpenShift, Docker, GitHub, GitLab, CircleCI, Jenkins.

- [ ] **Build Tools**
   Maven and Gradle configurations have equivalent capabilities to build for production and run in dev mode.

**Requirements**
1. Vaadin platform artifact `vaadin-dev` becomes `optional=true` and `scope=test` so it is not by default included into runtime, but added when app is running from a test scope.
2. Vaadin ensures that having two same `Application` Spring-Boot Java classes in a single application do not trigger a "Multiple classes implementing `AppShellConfigurator` were found" errors automatically or by requiring an opt-in flag.
3. Vaadin does not leave a project with inconsistent configuration when jumping between production builds and dev runs. E.g. I was caught by "no DevModeHandlerManager implementation found but dev server enabled" error when tried to run with `mvn spring-boot:run` after production build.
4. We recommend running `mvn package quarkus:dev` for dev runs, but this would clash with a production build as that also runs `mvn package`. This is to be re-considered why we have `package` here and how we can avoid it.
5. Vaadin maven and gradle plugins should detect CI/CD build environments by scanning documented ENV variables. This scanning is enabled by default, but can be disabled by a parameter.
6. start.vaadin.com , start.spring.io and our standalone starter repos should be updated to this new approach.

**Other considerations**

1. We can delete the `prepare-frontend` goal from Vaadin plugins as now the production build trigger is not a plugin's goal, but the way how you run the app.

**Comment by knoobie** [comment]
What happens with the spring boot run goal? e.g. starting the application via IDEA's Maven Plugin Overview?

<img width="313" height="290" alt="Image" src="https://github.com/user-attachments/assets/0852ebb3-a748-4859-bb79-2927a5ef4474" />

**Comment by Artur-** [comment]
At least there is no use case that involves using `mvn spring-boot:run` to run in real production. As always, it should be avoided in the IDE though as it breaks debug functionality in some IDEs

**Comment by mshabarov** [comment]
I assumed `mvn spring-boot:run` would build and run the app in production mode.

**Comment by Legioth** [comment]
`mvn spring-boot:run` without `package` may outright fail or give inconsistent results. You're supposed to use `mvn spring-boot:test-run` from that list for launching from the IDE.

We should probably make sure we show a reasonable error message if doing `mvn spring-boot` run without `package` before, though it might be difficult to notice if the result of a previous outdated `package` run is still present in `target/`.

**Comment by Artur-** [comment]
Are you saying that `mvn package spring-boot:run` is a way to test the production version? How would you test it from the IDE?

**Comment by knoobie** [comment]
> You're supposed to use mvn spring-boot:test-run from that list for launching from the IDE.

Where is this written? https://docs.spring.io/spring-boot/reference/using/running-your-application.html has no mention of test-run at all. Even here.. it calls "test-run"  *sometimes* ( "Sometimes it is useful to run a test variant of your application. " https://docs.spring.io/spring-boot/maven-plugin/run.html)

I know a lot of people that use `mvn spring-boot:run` for years in their IDEs.

**Comment by Legioth** [comment]
If you're a Spring Boot user and want to [use Testcontainers](https://docs.spring.io/spring-boot/reference/features/dev-services.html#features.dev-services.testcontainers) in dev mode, then you use `spring-boot:test-run`. If you're a Vaadin user and want to use Vaadin's dev mode, then you do likewise.

The key reason for why this is needed for Vaadin's dev mode is that you want to have additional dependencies on the classpath that shouldn't be there when you do `mvn package`. The only sensible way that we have found for that is defining them with `<scope>test</scope>` and then running with `mvn spring-boot:test-run`.

If someone has alternative ideas for how to deal with that, then it's not yet too late to come up with an alternative plan. But without other alternatives, we're assuming that it's an acceptable tradeoff that developers will have to learn to use `test-run`. We should make sure there's some guidance in place to reduce the risk of confusion. One thing that comes to mind is to integrate with the IDE detection to show some kind of warning in case you're by accident running in production mode from your IDE.

**Comment by knoobie** [comment]
Sounds like the perfect candidate for an RFC :) I doubt a lot of people are going following this issue

**Comment by simasch** [comment]
Please don't mix `spring-boot:test-run` and `spring-boot:run`.

Those are two different use cases:

**`spring-boot:test-run` or starting the TestApplication class from the IDE**
I want to have the test dependencies for running the application (e.g. Testcontainers)

**`spring-boot:run` or starting the Applicaiton class from the IDE**
I want to have the "real" classpath for running the application.

In my applications, I use both during development!

Either I want to have a database in Testcontainers or I want a local or remote test database with a specific test data set
 
**I vote strongly against your plans!**

**Comment by simasch** [comment]
> `mvn spring-boot:run` without `package` may outright fail or give inconsistent results. You're supposed to use `mvn spring-boot:test-run` from that list for launching from the IDE.
> 
> We should probably make sure we show a reasonable error message if doing `mvn spring-boot` run without `package` before, though it might be difficult to notice if the result of a previous outdated `package` run is still present in `target/`.

That's a wrong assumption. **spring-boot:run runs before package phase!** And the jar is NOT in the classpath.

It even goes further if you run mvn validate Spring Boot adds the output directory to the class path instead of the built jar!!!

From the Spring Boot parent POM:

```
       <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-failsafe-plugin</artifactId>
          <executions>
            <execution>
              <goals>
                <goal>integration-test</goal>
                <goal>verify</goal>
              </goals>
            </execution>
          </executions>
          <configuration>
            <classesDirectory>${project.build.outputDirectory}</classesDirectory>
          </configuration>
        </plugin>
```

The problem is that you cannot run integration tests with the Spring Boot JAR as it is not a standard Java JAR.

**Comment by simasch** [comment]
Why not run the frontend build in the pre-package phase instead of the compile phase?
This is how Spring Boot creates the uber JAR.

**Comment by Legioth** [comment]
> That's a wrong assumption. spring-boot:run runs before package phase!

Not sure what you mean. If I run `mvn package spring-boot:run` on the command line, then it runs e.g. `surefire:3.2.5:test`, `jar:3.4.2:jar` and `spring-boot:3.3.2:repackage` _before_ running  `spring-boot:3.3.2:run`. If I run only `mvn spring-boot:run`, then it runs none of those.

> Either I want to have a database in Testcontainers or I want a local or remote test database with a specific test data set

Do you mean that you sometimes want to run Vaadin in development mode with a Testcontainers database and other times run Vaadin in development mode with a regular database?

> Why not run the frontend build in the pre-package phase instead of the compile phase?

The big challenge is to make `mvn package` create something that doesn't contain the `vaadin-dev` dependency while still also having a simple and reasonably standard way of running Vaadin in development mode with `vaadin-dev` on the classpath for the regular development mode experience.

**Comment by simasch** [comment]
Why would you run mvn package spring-boot:run?

mvn spring-boot:run uses the classpath from the output directory, not the uber JAR, which can't be used as it's not a regular JAR.

> Do you mean that you sometimes want to run Vaadin in development mode with a Testcontainers database and other times run Vaadin in development mode with a regular database?

Yes, we do that in most projects.

> The big challenge is to make mvn package create something that doesn't contain the vaadin-dev dependency while still also having a simple and reasonably standard way of running Vaadin in development mode with vaadin-dev on the classpath for the regular development mode experience.

The problem is that Vaadin uses transitive dependencies. If we were to add vaadin-dev as a separate dependency with scope runtime, like Spring Boot does with the Dev tools, the problem would be solved.

**Comment by simasch** [comment]
Ah I see there must be some tricks with the Spring Boot dev tools that they don't get into the uber JAR 😔

**Comment by simasch** [comment]
Indeed they have a flag on the Spring Boot Maven Plugin to exclude the devtools. WTF

**Comment by Legioth** [comment]
Exactly. The Spring Boot Maven plugin has hardcoded logic specifically to exclude their own dev tools dependency. There's no obvious way of getting the same treatment for our dev dependency even if we wouldn't make it non-transient. That's why we see the test classpath as the least bad alternative.

I wonder if it would be reasonable to have something custom based on profiles for the intersection where Testcontainers and Vaadin dev dependencies are both used and you want to toggle each independent of the other?

**Comment by simasch** [comment]
I don't have any problem with the production profile. I like it as it is. I usually have more profiles, like an IT and a Coverage profile. 

It's important not to change the way it is. Otherwise, I will have trouble with most of my projects. This would be a huge breaking change.

**Comment by simasch** [comment]
Btw. just changing something because some people don't understand Maven is not the way to go

**Comment by knoobie** [comment]
Defining the problem with other words: 
- Vaadin should build itself automatically "production ready" if no Vaadin Development Tools are found(?) 
- Vaadin Development Tools should be either declared developmentOnly (Gradle) or optional=true/scope=runtime and Spring Boot still includes it, even tho it should not. 

Wouldn't all this be solved if you could get Boot on board? To either get it fixed or allow your devtools to be added to the list of exclusion / create a hook for it?

**Comment by Dudeplayz** [comment]
I think the simplest and cleanest option would be to request a change at Spring Boot to support some way to work like their dev-tools or to add some goal like quarkus:dev. I think other software and tooling would also profit from it.

**Comment by ThomasVitale** [comment]
Spring Boot's way of supporting Testcontainers at development time presents some challenges (I shared my thoughts in  https://github.com/spring-projects/spring-boot/issues/46367). I wouldn't recommend it for Vaadin. Even though it's currently how that works in Spring Boot, it's worth noting that many projects don't use the Spring Boot Testcontainers support at development time in favour of the Spring Boot Docker Compose support, which runs from the main classpath. For those projects, having Vaadin requiring starting the app from the test classpath would mean changing the development workflow and introducing a non-standard way of working with a Java application in a local environment.

The underlying problem is that Spring Boot doesn't support defining dev-only dependencies. For Gradle projects, that is mitigated by the custom configurations added by the Spring Boot Gradle Plugin (`developmentOnly` and `testAndDevelopmentOnly`). For Maven projects, that's not available since Maven doesn't allow that type of customizations. The Spring Boot Maven Plugin takes care of removing the Spring Boot Docker Compose and Spring Boot DevTools from the production artefact, but it's not a feature available for libraries to hook into (there are a few issues about this problem, all linked under https://github.com/spring-projects/spring-boot/issues/46817).

Given that, I totally understand why the suggestion for Vaadin to adopt the test-classpath requirement to run the app during development. It would be a much better developer experience not having to configure and enable the Maven profiles. But doing that while still maintaining the current development workflows is not trivial. And I'm not sure there are many alternatives, besides developing a custom plugin trying to extend the Spring Boot Maven Plugin to deal with excluding dev-only dependencies.

It would be great if Spring Boot provided support for dev-only or dev/test-only dependencies also in Maven. Also, as a side improvement, it would be great to have a programmatic way to determine the app bootstrap mode (dev, test, prod: https://github.com/spring-projects/spring-boot/issues/46368).

**Comment by Legioth** [comment]
The main goal is indeed to make `mvn package` do what people expect it to do: create an artefact that works well in production. It's not that they don't understand how Maven works but rather that they end up wasting time and building up frustration by looking into other directions before they come to think of Maven profiles. 

The optimal solution would indeed be that the mechanism used for Spring Boot devtools would also be available for 3rd party dev dependencies. There are multiple tickets in that area that have been open for quite some time so I'm not very hopeful about that being resolved any time soon. 

If we proceed with the test classpath approach, then the main breaking change would be that `vaadin-dev` is no longer a transitive dependency. I assume it would be a relatively simple step to adjust `pom.xml` to compensate for that when upgrading to Vaadin 25 if you want to keep using the profile, even though I haven't yet seen a practical example for that.

Now when I think about it, we might even keep it as a transitive dependency in `com.vaadin:vaadin` if we also proceed with the new optimized starter dependencies described in https://github.com/vaadin/platform/issues/5230.

**Comment by simasch** [comment]
I'm with @ThomasVitale and wouldn't use the test classpath for development by default because this would be specific to Vaadin (as the production profile is), and you would have to explain to the developer why to run the TestApplication or spring-boot:test-run, even if they are not using Testcontainers. 

There may be even other side effects depending on what the developers add as test dependencies.

IMO, this is the bigger hack than having the production profile.

**Comment by Legioth** [comment]
> because this would be specific to Vaadin (as the production profile is)

I think this points to a new way of looking at this case even though it's not directly spelled out. With the mechanisms currently available for Spring Boot, we must have something **specific to Vaadin at one point or the other** – either for dev mode or for production mode. Once you've gotten used to doing that specific thing, then it comes relatively natural and you're familiar with the symptoms of occasionally forgetting so that you can easily recover. The problem is for new Vaadin users who are not yet used to the Vaadin specifics.

Assuming either specific is of roughly the same magnitude (which I claim they are in this case), then we should prefer the approach that shifts the **need to learn about specifics later in the learning journey**. You would typically want to run a Vaadin application in development mode before you run it in production mode which means that it's more important that **nothing specific is needed for development mode**.

We do know that some users currently struggle with production because they have missed the instructions about using a custom Maven profile. It's fair to assume that those same users would also miss the instructions that tell them to use `test-run` or the alternative `main` method for development mode. And struggling at that phase would be even more problematic than struggling with production deployment because you have less motivation to find a solution if you haven't had some initial success with development mode.

There is also the argument that having the profile defined in an otherwise quite simple `pom.xml` gives the impression that there's something complicated about running or building Vaadin applications. While this is to some extent true, there's also the positive aspect that it provides a way of discovering that the profile exists. With the alternative approach, the hints about having to run development mode in a custom way aren't as easily discovered and thus even more likely to cause new users to struggle.

Based on this new perspective, I'm leaning towards **keeping the production profile and not using the test classpath for development mode**. This doesn't mean that it's set in stone that we will never consider the test classpath, but rather that some new fact or perspective would be needed for it to be considered again.

**Comment by Legioth** [comment]
And thank you to everyone for providing their input so far! In particular, the very [clear analysis](https://github.com/vaadin/flow/issues/17737#issuecomment-3266723455) by @ThomasVitale was very helpful in seeing both the pros and cons of the suggestion. :+1:

**Comment by mstahv** [comment]
I went for a holiday and suddenly there a lot of action on this very important issue that I already thought will never be taken seriously 👍 I'm not too happy though about your resolution (do nothing). If one follows new Vaadin users struggles, one quickly sees this is one of the top issues people have, and I'm quite sure many evaloators currenlty drop off as they can't deploy even a simple app. And also users with bit more Vaadin experience, like me, face this regularly. I had also learned to live with the inconveniece, but nowadays, with my custom project setup, I always have a wide smile on my face when pushing sources of my examples/demos to my private PaaS server without special configurations (and things just work like they would with a regular cloud native Spring Boot app).

I agree it would be the perfect solution if Spring Boot would just fix itself (like make optional dependencies work as one would expect, like they handle their devtools). But given the history on those SB issues, I'm quite pessimistic about getting anything forward on that front anytime soon. Thus I still believe a separate "DevModeApp" approach would be the best compromise we could take today for Spring Boot, but not the only way to make things better (if there are some actual blockers on that approach). For that approach to work we would naturally need (didn't check the plan if these were covered):

 * Proper modularisation (as discussed in separate issue) in a backwards compatible manner (and this way keep production profile functional for those who currently have it).
 * Improved logging in production mode to guide people to dev mode
 * IDE plugin/doc changes (e.g. so that Vaadin IDEA plugin could help to guide people to right direction)
 * Mechanisms (profile and/or flags) to explicitly choose production mode or development mode for integration tests, even if development tooling is on classpath.
 
@simasch, I probably missed something regarding the Testcontainer related concerns, how would this Vaadin change affect that? (given that choosing between production/development was still possible somehow, as it should be, but just the default for new projects would change)

Related to this issue, I had a great success motivating @mcollovati  to prototype a bit on this topic using Quarkus. To me it looks like on that front we have a chance to make things in a "perfect way" already today 🤔 This is how we can do it on that front:

 * **Quarkus tooling already has a  "dev mode app"** built into the tooling and most(all🤔) developers use that to start the app while developing (`quarkus dev` or launching that via IDE plugin). So change at all needed for current workflow. (Irrelevant personal preference: I like SB approach to just launch a main method from IDE more)
 * **Quarkus extensions can drop in custom dev dependencies and tasks to dev mode**, so we don't need to pollute the default classpath with our (rather massive) dev mode, but dev mode is still there when people develop. The default build artifacts stays clean and production ready (deploy like regular Quarkus app, no Vaadin specifics needed), without custom "vaadinisms".
 * **No need for Vaadin specific profiles at all** (although there could be some custom needs, e.g. like what @simasch was doing with his SB project). Implementation detail in the PoC: application.properties has vaadin.build.enabled=true flag, should be implicitly on in final solution, didn't check how Marco did it for now.
 * The cool part is that we **don't really need a separate Vaadin Maven plugin at all** and "Vaadin dirt" in the build file could ideally become just a single dependency. 😎 (The extention also detects a "regular build" and then makes the production bundle with the same tooling used in our Maven/Gradle plugins).

Based on my quick testing during the weekends rainy moments, I already mentioned to Marco that **this is the way Vaadin should work**. If you want to try and play with this PoC (and please do!):

 * Check out spikes/embedded-plugin branch and build local snapshot https://github.com/vaadin/quarkus/tree/spikes/embedded-plugin
 * Test e.g. using spikes/embedded-plugin from the example/starter: https://github.com/vaadin/base-starter-flow-quarkus/tree/spikes/embedded-plugin quarkus dev -> Vaadin starts in dev mode, mvn install -> prod mode, mvn dependeency:tree looks a lot less scary than with "typical Vaadin project" (half contributed to the fact that there is no Hilla and its transitive dependecies, but also the dev mode "dirt" is not listed).

**Comment by Legioth** [comment]
I 100% agree that the Quarkus solution is a very good one. It works thanks to the way Quarkus users from the very start have learned to launch using `quarkus:dev` or equivalent. Unfortunately, the default behavior for Spring Boot users is either `spring-boot:run` or launching `Application.main` through the IDE and we do not have the same mechanisms there that are available to Quarkus extensions.

> I'm quite sure many evaloators currenlty drop off as they can't deploy even a simple app.

I'm quite sure that even more evaluators would drop off even before they get to the point where they want to deploy something with the opposite setup, i.e. that building for production would become as easy as running dev is now whereas running dev would become as difficult as building for production is now.

**Comment by mstahv** [comment]
In my PoC (with dev mode app in test sources), I kill the app with a message that guide towards the dev mode if started directly without packaging. It ought to be pretty easy to figureout the right way then for those who don't read instructions 🤔 

One can of course also claim that it already happens for invalid production deployment, but we have quite a lot of proof points that those are hard to read even for experienced Java users (who knows how to read the logs in Beanstalk anyways 🤷‍♂️). We (me including) have downplayed this issue in the past with RTFM comment, but my current assumption is that this comes quite early in the evaluation (deploy already the hello world to some PaaS, local server or test with some CI server that deployment works). Maybe in this container era just building the image and testing it runs is enough, but that sure is broken by default ATM.

**Comment by mstahv** [comment]
The nasty thing is that it is almost impossible to get "numbers" for this kind of decissions. Only somewhat bullet proof method to test this kind of experience would be to build build rather polished PoCs and then hire a bunch of Java developers to test (unaware of Vaadin before hand) and follow analytically their evaluation.

**Comment by simasch** [comment]
The common way to run a Spring Boot application during development is either spring-boot:run or starting the main method of the Application class.

If you don't choose the Testcontainers dependency on https://start.spring.io there will be NO TestApplication! 

Using the test classpath during development time is a nasty hack and can cause side effects depending on the test dependencies you configure, and you risk having different behavior than in production.

I have many customers not using Testcontainers, and for them, this would be a huge breaking change.

I still advocate a strict NO for this change!

**Comment by mstahv** [comment]
I have understood the drawbacks regarding our "old habbits" and that we are kind of "abusing" test scope of dependencies, but wanted to understand how your needs discussed in https://github.com/vaadin/flow/issues/17737#issuecomment-3266251054 would get worse.

If "done right", this would NOT be a breaking change. Existing apps and users should not need to change anything. We should just get it into new projects (and be very open to roll back if we get hints that there are issues, especially among NEW users). 

Using test classpath indeed is a hack, just like when used with testcontainers & spring boot (for development), but I don't see that as that big problem. At least I don't see it very different from the current where Vaadin development tooling is on the classpath both while developing and testing, but NOT when running in production. Dependency conflicts I have faced too often with Vaadin, but they have been typically because of other runtime dependencies and dependencies brought in via our development mode/license checker/hilla. Those might get slightly better as regular compilation (during development) would have bit less stuff on the classpath 🤔

**Comment by Legioth** [comment]
We seem to somehow talk past each other and I'm not sure what is the cause of that. So let's try to make my perspective really explicit so that you can point to the exact aspect that you see differently.

The relevant parts of the typical evaluator journey can be simplified as only two steps:
1. Create something cool locally using development mode
2. Create a production build and deploy it to share the cool stuff with others

Due to limitations in what Spring supports, we have to accept some friction in one of the steps and we thus have two options.

**Option 1: status quo**
1. Works as expected
2. You need one unexpected extra detail and you can get a helpful error message if you don't (the symptoms you get are not optimal today but that can be improved)

**Option 2: suggested approach with the test classpath**
1. You need one unexpected extra detail and you can get a helpful error message if you don't
2. Works as expected

One key aspect where we might see things differently is in my assumption that the friction in each case is roughly of the same magnitude. The way I see it is that both are about one "bit" of extra things needed e.g. adding `-Pproduction` or changing `:run` to `:test-run`. Our opportunities to detect the situation and provide a helpful message are also roughly equivalent in both cases.

Based on this equivalence, I'm making the assumption that it's better to have the friction later in the process. With Option 1, you already have one significant success from the first step on your mind when you encounter the friction and this gives you additional motivation to push through to find a solution. Option 2 on the other hand gives you friction before you have had any chance to really succeed with anything which also means that you have lower motivation to give it a second chance.

Where do our mental models diverge?

**Comment by mstahv** [comment]
The current compromise absolute has a slightly better early experience for complete newbies (who are new to Java, Maven, Vaadin and Spring Boot), on their initial experience. As this group also actually reads instructions, and if we have keep them on our instructions on the deployment phase too (meaning they pick the hosting solution we hint), we can probably keep them happy while deploying too. Keeping things as they have been is also always a safer for old dogs who have already learned "the Vaadin way" (like me until I started investigating other approaches with these half baked PoCs that I now use in my own projects).

But I'm not convinced this is now good for experienced Java developers/architects "who know what they are doign". This weirdness, that we cause by not creating a production ready artifact by default (with `mvn package` or `mvn spring-boot:build-image`, `cf push` or `pack build my-image` done behind the scenes by PaaS/CD pipeline) and then fail the deployment later with crycptic errors, causes big WTFs and if they end up reading the instructions it forces them to struggle with custom deployment and build pipeline settings. Several this kind of evaluators have come directly to their contacts in Vaadin to ask help for this issue. This group is our most valuable evaluators, their suggestions are listened by the wider audience. Their experience is the most important.  I'm confident we have lost many potential "Vaadin advocates" because of this oddity. Days for me since explaining this weirdness is currently 0.

As discussed, these same valuable "Java experts" are likely to investigate "what has this thing eaten" with `mvn dependency:tree`. With the starndard "production first" approach they would get a lot less frightening listing.

This is why I would be ready to add a small hurdle for the early evaluation, if at the same time we can make Vaadin behave more like other tooling out there. If utilizing the test scope for dev mode enhancers is considerd too alien, I might be ready to go so far that instead of "production" profile we would have a dev profile (which we'd turn on by IDE plugin, instructions or with some other not-yet-discovered trick), like it was planned for basic war projects in the Mikhail's draft.

I have stopped investing alternative options as I have been so happy with the Spring Boot+Testcontainers (for development RDBMS connection) inspired approach in my examples and hobby related "production apps", but there sure could be other ways to solve this. I might be even eagerly continuing on finding alternative ways, but it is not meaningful if I don't understand all concerns around this. *People will try src/main/java first* is well understood.

BTW. Even having a dev profile (instead of production profile) that would be on by default would be better than what we have today (exclusions in our current production profile "smell"), but naturally it wouldn't fix the problems expressed in this ticket and requires getting our currently messy module structure cleaned up.

Technical not about your "options" and especially the "step 1". It already kind of has the assumption that the evaluator knows what is the "developement mode" in Vaadin. Without it one could claim that in step 1 in option 1 Vaadin currently works unexpectedly. There is for example "weird logo on the lower right corner" (dev mode button) and starting the server process takes ages (compared to server startup of Vaadin with proper production build).

**Comment by simasch** [comment]
When reading your comment @mstahv it makes the impression that Vaadin is the only framework in the world that has an extra step for a production build.

I also use JTE with Spring Boot, and because JTE templates are pre-compiled, I also need this step: https://jte.gg/spring-boot-starter-3/
Which is the same as we had decades ago if we want to have precompiled JSPs.

I have many customers using Anguar. Some of them also have a single deployment JAR, and they need a production build for the JavaScript part.

**Comment by mstahv** [comment]
Vaadin will always have a production build, at least if we want to keep the client side extensible and still compact. Quickly looking the  JTE code-generation/compilation is bit of the same. The main point is just what is the default.

I followed the linked page to test JTE (for the very first time ever) to get some insights. After some fiddling (adding maven plugin and figuring out how to name the template file etc) console showed me this looong exception on a single line, from its end I got (same on both IDE main method start and mvn packge +  java -jar.):

```
Factory method 'jteTemplateEngine' threw exception with message: You need to either set gg.jte.usePrecompiledTemplates or gg.jte.developmentMode to true
```

At least they make good job at people realising how it works. I took the first option (added `gg.jte.usePrecompiledTemplates=true` to application.properties) (and probably would have taken as an evaluator first too) and then everything works as expected with launching from IDE, java -jar and with mvn spring-boot:build-image. I can only wonder why usePrecompiledTemplates is not the default (at least on Spring Boot) as they anyway do the code generation/compilation during the build with maven plugin, it is just some flag that prevents using those precompiled classes. Or it could use those implicitly if they are there and dev mode is not on 🤔. Maybe this has been intentional decision that they want teach evaluators about the dev mode as early as possible 🤷‍♂️

I naturally wanted to test the development mode as well. In the past, I probably would have either kep their developmenMode configuration as a non-committed change in my local development environment, in `applicaito.propeties`. And after I first time accidentally commit it (and break the live production environment 😂), I'd move it to a JVM argument in my IDE lanch config (screenshot). Now that I'm accustomed to having a separate DevModeApplication in the test sources (thanks to Testcontainers), I'd probably adapt a similar approach for JTE development too.

One note about JTEs difference to Vaadin dev mode: their dev mode, and library in general, is lightweight (measured in kilobytes), and it pulls in no external dependencies. Based on my with my quick learning, there don't seem to be an easy way to get the "dev mode" out of the production builds at all, there it is just some configs toggling it on/of explicitly.

BTW. That project for sure hasn't paid as much attention to DX as we do (although we are probably just as blind to similar hindurances on our side). For example I had to disable the precompilation explicitly (instead of devMode doing that for me implicitly). I also wonder why they have spring boot starter that pulls in `gg.jte:jte`, but they still instruct have have that separately as a dependency 🤷‍♂️ And they don't mention `spring-boot-starter-web(flux)`(Spring MVC or WebFlux) as required dependency as you really don't get far (in Spring Boot environment) without either. Anyways, based on my experience, let's not copy-cat their DX. If you @simasch happen to know the maintainer, don't hesitate to pass my feedback from this "industry espionage" forward 🧸
<img width="806" height="680" alt="Image" src="https://github.com/user-attachments/assets/dd821b01-ec67-4624-9cce-7a436f678163" />

**Comment by knoobie** [comment]
+ an "optional" fix for boot is on the way for 3.5 as flag 🥳

https://github.com/spring-projects/spring-boot/commit/f3b44031a671c54d2c88bc974341f79ca30066ad

+ a way to define a dependency as development tool to be released for 4.0 🥳

https://github.com/spring-projects/spring-boot/commit/27715dee202af68dbb0d559196ab4a438d90aafb

**Comment by mstahv** [comment]
Whining feels bad but apparently sometimes doing it long enough makes a difference 😎 Please all continue that on our issues too, I know we often manage to ignore similar and obvious easy fixes on our side too 🧸 There is just so much noice on big projects.

**Comment by mstahv** [comment]
There was some hickup with SB 3.5.x snapshot build, but built it locally and played a bit with my PoCs, this is now the desired outcome:

https://github.com/viritin/viritin-sb/tree/with-sb-optional-fix/example-flattened-flow-only

In 4.0, the extra config of SB plugin can be removed (and hopefully we could in the future do with just one goal).

Looks very promising, almost as perfect as with Quarkus PoC. I'll test with couple of "real" projects on a Packeto buildpacks based PaaS once a new snapshot is published in SB repo.

Now I wonder if one can now also configure the Testcontainers for development in this manner, by adding an optional dependecy containinng configurations for test RDBMS container 🤓 Well, I have personally already accustomed to use those separate classes so I'll let somebody else investigate that one.

**Comment by mshabarov** [comment]
Two things to check and consider before starting the implementation:
- Does the new `development-tool` marker in Spring-Boot work for transitive dependencies as well, i.e. do we need to mark all our dev dependencies with it, not only `vaadin-dev`;
- If we finally need to mark `vaadin-dev` as optional, this should be done only in vaadin-spring-boot-starter to limit this only to Spring-Boot projects, otherwise it would need changes in non-Spring projects and be a breaking change.

**Comment by tltv** [comment]
We don't need to introduce changes in `vaadin-spring-boot-starter` to make `vaadin-dev` optional by default. It's enough to use this new `includeOptional=false` SB feature in application's pom.xml and add `vaadin-dev` dependency explicitly as an optional. This seems to be enough to exclude `vaadin-dev` when building or running with maven, but including it when run via IDE. No breaking changes needed then.

With this approach, and accepting that running with maven will always run production package, there's no need for production maven profile e.g. in Vaadin starters.

However, if `mvn package` should always produce production package and `spring-boot:run` should run with Vaadin dev tools, there might be need for maven profiles since `build-frontend` goal need to be skipped somehow for `spring-boot:run`, and profiles with `activation` makes that possible together with a system property.

**Comment by Legioth** [comment]
> since `build-frontend` goal need to be skipped somehow for `spring-boot:run`

Or then we just change it from `<phase>compile</phase>` to e.g. `<phase>package</phase>`? If we still need something to detect the root of the Maven project, for the cases where the current working directory isn't correct, then we can re-introduce a separate goal for that instead of using `build-frontend`?

Edit: Oh, there's already `prepare-frontend` for the path setup so then it might be enough to just connect `build-frontend` to `package` instead of `compile` and move it out from the profile?

**Comment by tltv** [comment]
Combining `build-frontend` goal to `prepare-package` phase allows `spring-boot:run` also run in development mode without any profiles. Minimum configuration looks like this:

```xml
<dependency>
    <groupId>com.vaadin</groupId>
    <artifactId>vaadin-dev</artifactId>
    <!-- optional=true is needed together with includeOptional in spring-boot-maven-plugin. -->
    <optional>true</optional>
</dependency>
```

```xml
<build>
	<defaultGoal>spring-boot:run</defaultGoal>
	<plugins>
		<plugin>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-maven-plugin</artifactId>
			<configuration>
				<!-- New in Spring Boot 4. Set false to exclude optional dependencies like vaadin-dev. -->
				<includeOptional>false</includeOptional>
			</configuration>
		</plugin>
		<plugin>
			<groupId>com.vaadin</groupId>
			<artifactId>vaadin-maven-plugin</artifactId>
			<version>${vaadin.version}</version>
			<executions>
				<execution>
					<goals>
						<goal>prepare-frontend</goal>
					</goals>
				</execution>
				<execution>
					<!-- Package for production -->
					<goals>
						<goal>build-frontend</goal>
					</goals>
					<phase>prepare-package</phase>
				</execution>
			</executions>
		</plugin>
	</plugins>
</build>
```

**Comment by mstahv** [comment]
So it is now starting to look like https://github.com/viritin/viritin-sb/blob/with-sb-optional-fix/example-flattened-flow-only/pom.xml 🤔

**Comment by mstahv** [comment]
For **new** basic war/jakartaee projects I suggest to invert the setup (defaulting to production mode there as well, but "development" profile that includes our dev tooling). Then the awesome setup that Marco hacked for Quarkus and we are good to go?

The only thing where it might backstep (that I know) is then if one wants to do "integration tests" in production mode with @SpringBootTest, without launching the actual artifact. I'm not sure how common or important that is (as if testing without vaadin-dev is important, then I'd expect a proper IT test that really launches the artifac is desired to exclude other test/dev artifacts too), but I did some local test to verify it can be done with profiles if somebody really needs it. Anyways, this kind of setup would become a bit easier if there was an explicit flag to disable development mode, even if the artifact would be on the classpath (I'd imaging this to be quite easy to do, IIRC we even had this in the past, or maybe even still do 🤔).

**Comment by Legioth** [comment]
> For new basic war/jakartaee projects I suggest to invert the setup (defaulting to production mode there as well, but "development" profile that includes our dev tooling).

I'd do the opposite and just leave it as it is. We're not optimizing for those cases anyways so then it's just as good to not make any changes so that old documentation remains valid.

**Comment by mshabarov** [comment]
Closing this as done.

**Comment by mstahv** [comment]
Gave the changes a try, overall looks good. I'd hope though that for exampel @simasch would test the current beta release as you had some concerns I didn't really understand properly.

I think I found bit weird/surprising was that vaadin-dev still comes via vaadin-core. It is exlicitly defined optional in the starter now to make this all work, but needed to remove that when ensuring one can still "test in production(mode)", and then again ended up having those ugly exclusions. The whole setup was bit complicated in general, but not really sure how common it is (My current habit for E2E tests is to run them in development mode via SpringBootTest. If I would take things really serious, at least some integration tests would startup my server as in production, not via SpringBootTest annotation 🤷‍♂️).

**Comment by Artur-** [comment]
Wouldn't removing `vaadin-dev` from `vaadin-core` break all existing applications? Would that be worth it?

**Comment by simasch** [comment]
@mstahv 
I've tried 25.0.0-beta2 with Spring Boot 4.0.0-RC1, and with the optional dev dependency, nothing is affected, and it's also not required to run it in test mode.
So I'm happy with the result.

Great works and kudos to Spring Boot to also exclude other optional dependencies when doing the repackage.

**Comment by mstahv** [comment]
> Wouldn't removing `vaadin-dev` from `vaadin-core` break all existing applications? Would that be worth it?

vaadin-core is kind of obsolete in spring boot projects (brought in via starter, at least previously), one option would be to exclude it from the starter. But probably best would be to have a new one without vaadin-dev, with some name,  as I suggested in #5230.

**Comment by mshabarov** [comment]
> But probably best would be to have a new one without vaadin-dev, with some name, as I suggested in https://github.com/vaadin/flow/issues/5230.

Wrong link? 5230 is about completely different topic.

**Comment by mshabarov** [comment]
Reopened as this needs at least Vaadin project updates for Spring Initializer. And some research for Gradle-based projects (might be a separate feature though).
